xds/src/main/java/io/grpc/xds/PriorityLoadBalancer.java

```diff
@@ -0,0 +1,297 @@
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (ChildLbState child : children.values()) {
+      child.tearDown();
+    }
+  }
+
+  private void tryNextPriority(int startIdx) {
+    for (int i = startIdx; i < priorities.size(); i++) {
+      String priority = priorities.get(i);
+      if (!children.containsKey(priority)) {
+        ChildLbState child = new ChildLbState(priority);
+        children.put(priority, child);
+        child.updateResolvedAddresses();
+        return; // Give priority i time to connect.
+      }
+      ChildLbState child = children.get(priority);
+      child.reactivate();
+      if (child.connectivityState.equals(READY) || child.connectivityState.equals(IDLE)) {
+        logger.log(XdsLogLevel.DEBUG, "Shifted to priority {0}", priority);
+        currentPriority = priority;
+        updateOverallState(child.connectivityState, child.picker);
+        for (int j = i + 1; j < priorities.size(); j++) {
+          String p = priorities.get(j);
+          if (children.containsKey(p)) {
+            children.get(p).deactivate();
+          }
+        }
+        return;
+      }
+      if (child.failOverTimer != null && child.failOverTimer.isPending()) {
+        return; // Give priority i time to connect.
+      }
+    }
+    // TODO(zdapeng): Include error details of each priority.
+    logger.log(XdsLogLevel.DEBUG, "All priority failed");
+    updateOverallState(TRANSIENT_FAILURE, new ErrorPicker(Status.UNAVAILABLE));
+  }
+
+  private void updateOverallState(ConnectivityState state, SubchannelPicker picker) {
+    if (!state.equals(currentConnectivityState) || !picker.equals(currentPicker)) {
+      currentConnectivityState = state;
+      currentPicker = picker;
+      helper.updateBalancingState(state, picker);
+    }
+  }
+
+  private final class ChildLbState {
+    final String priority;
+    final ChildHelper childHelper;
+    final GracefulSwitchLoadBalancer lb;
+    // Timer to fail over to the next priority if not connected in 10 sec. Scheduled only once at
+    // child initialization.
+    final ScheduledHandle failOverTimer;
+    // Timer to delay shutdown and deletion of the priority. Scheduled whenever the child is
+    // deactivated.
+    @Nullable ScheduledHandle deletionTimer;
+    @Nullable String policy;
+    ConnectivityState connectivityState = CONNECTING;
+    SubchannelPicker picker = BUFFER_PICKER;
+
+    ChildLbState(final String priority) {
+      this.priority = priority;
+      childHelper = new ChildHelper();
+      lb = new GracefulSwitchLoadBalancer(childHelper);
+
+      class FailOverTask implements Runnable {
+        @Override
+        public void run() {
+          if (deletionTimer != null && deletionTimer.isPending()) {
+            // The child is deactivated.
+            return;
+          }
+          int idx = priorityToIndex.get(priority);
+          updateOverallState(
```

Comment:
Isn't this wrong? Failing over from one priority to another causes a TRANSIENT_FAILURE picker to the upstream policy (and maybe Channel)?

Reply:
I think when current priority fails explicitly, we should go to TRANSIENT_FAILURE to let the end user know something is already broken, before failing over to the next priority (which takes time to connect). I remember we had a discussion with @ejona86  before. An exceptional case might be the next priority is already READY but was in a previously de-activated state, in this case we can immediately re-activate the next priority and use its picker with out disruption.

Reply:
I think @markdroth had said the behavior should be to wait for the new priority to come online, like we do in startup. I can see it both ways, but delaying seems fair; that would be similar to our pick_first behavior. It is clear we would want this behavior to be the same cross-language.

Reply:
If the current priority fails and there is another priority to try, we should report CONNECTING.  I think we should report failure only if all priorities have failed.

In C-core, `TryNextPriority()` takes a boolean parameter called `report_connecting` that tells it whether to put the channel in CONNECTING when it finds a priority that needs time to connect:

https://github.com/grpc/grpc/blob/7f20bccd4ddeceaad0a8ad9532e68a0f9e10b359/src/core/ext/filters/client_channel/lb_policy/priority/priority.cc#L399

We call `TryNextPriority()` with `report_connecting=true` when the current priority fails and when we get our very first update after we are created.  Otherwise, we call it with `report_connecting=false`.

I've updated the psuedo-code in the design doc to clarify this.

Reply:
Change. Thanks for the pseudo-code update.